[jit] Implement ScriptProfile to collect instruction profiles.

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57398 [jit] Add Python API for ScriptProfile
* **#57397 [jit] Implement ScriptProfile to collect instruction profiles.**
* #57396 [jit] Add a utility class SourceRef to represent Source as keys

Summary:
Introduces two main classes in C++ runtime:

ScriptProfile is the implementation for enalbing and disabling interpreter
profiling in C++. This should be only used from Python, and we will add
corresponding Python API in the next diff.

InstructionSpan is a utility class to instrument execution of each single
instruction. A start timestamp is recorded in the consturctor, and an end
timestamp is recorded in the destructor. During destruction, this will send
runtime data to all enabled ScriptProfile instances.

Test Plan:
build/bin/test_jit --gtest_filter='ScriptProfileTest.Basic'

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133579](https://our.internmc.facebook.com/intern/diff/D28133579)

zhxchen17 committed:
[jit] Implement ScriptProfile to collect instruction profiles.

Summary:
Introduces two main classes in C++ runtime:

ScriptProfile is the implementation for enalbing and disabling interpreter
profiling in C++. This should be only used from Python, and we will add
corresponding Python API in the next diff.

InstructionSpan is a utility class to instrument execution of each single
instruction. A start timestamp is recorded in the consturctor, and an end
timestamp is recorded in the destructor. During destruction, this will send
runtime data to all enabled ScriptProfile instances.

Test Plan:
build/bin/test_jit --gtest_filter='ScriptProfileTest.Basic'

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit] Implement ScriptProfile to collect instruction profiles."


Summary:
Introduces two main classes in C++ runtime:

ScriptProfile is the implementation for enalbing and disabling interpreter
profiling in C++. This should be only used from Python, and we will add
corresponding Python API in the next diff.

InstructionSpan is a utility class to instrument execution of each single
instruction. A start timestamp is recorded in the consturctor, and an end
timestamp is recorded in the destructor. During destruction, this will send
runtime data to all enabled ScriptProfile instances.

Test Plan:
build/bin/test_jit --gtest_filter='ScriptProfileTest.Basic'

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

nikithamalgifb commented on torch/csrc/jit/runtime/script_profile.cpp:
```diff
@@ -0,0 +1,124 @@
+  datapoints_.push_back(std::move(datapoint));
+}
+
+const ScriptProfile::Stats& ScriptProfile::dumpStats() {
```
Can you add some comments for all the API's on what they do?

nikithamalgifb commented on torch/csrc/jit/ir/scope.cpp:
```diff
@@ -124,15 +124,16 @@ std::vector<InlinedCallStackEntry> InlinedCallStack::vec() {
   return r;
 }
 
+InlinedCallStackEntry InlinedCallStack::getEntry() const {
```
Can you add some comments to this too?

zhxchen17 replied:
oops, these were old code that isn't needed anymore, so nice catch!

nikithamalgifb commented on test/cpp/jit/test_script_profile.cpp:
```diff
@@ -0,0 +1,42 @@
+  EXPECT_EQ(stat.count, 1);
+}
+} // namespace jit
+} // namespace torch
```
Are these the only tests? Are you planning to add any Python based tests?

zhxchen17 replied:
yes, as the summary mentioned we will add Python API in the next diff, so right now we only test C++ parts.

I can add another test for some illegal sequencing of C++ API, though.

nikithamalgifb commented on torch/csrc/jit/runtime/script_profile.cpp:
```diff
@@ -0,0 +1,124 @@
+
+ProfilesRegistry& getProfilesRegistry() {
+  static auto& registry = *new ProfilesRegistry{};
+  return registry;
```
Do not forget to address the clang-tidy errors :)

zhxchen17 replied:
I think clang-tidy is warning about a separate issue, which is exposing this static variable to be globally accessible. (and I don't know how to fix this quickly and elegantly, few options: 1. encapsulate this in a LeakySingleton class, like folly (https://github.com/facebook/folly/blob/master/folly/Singleton.h#L720), 2. bypass lint error 3. expose this in a private static class function, and access it with friend functions. probably 3. feels better)

re `new`ing the object, returning static variable by reference in theory isn't inherently safer, because something could potentially call into the singleton after its destruction, in other words, the leaking is intentional. Since this will be used in Python, things can be different, but still it's always better to be cautious with global destruction order. 

If we want to save one allocation here, we can implement something like folly::Indestructible (https://github.com/facebook/folly/blob/master/folly/Indestructible.h) but that requires separate effort and isn't so different using `new` essentially. 

We can also talk about this during sync up :)

gmagogsfm commented on torch/csrc/jit/runtime/script_profile.cpp:
```diff
@@ -0,0 +1,124 @@
+
+InstructionSpan::InstructionSpan(Node& node) : datapoint_(node.sourceRange()) {}
+
+InstructionSpan::InstructionSpan(InstructionSpan&& other)
+    : empty_(other.empty_), datapoint_(std::move(other.datapoint_)) {
+  other.empty_ = true;
+}
```
Is this for moving `InstructionSpan`? Could we disallow moving to save the overhead of tracking/checking `empty_`? It would also reduce the cognitive load of knowing `c10::nullopt` is same as `InstructionSpan{empty_==False}`.

zhxchen17 replied:
yep, initially I thought the moving was unavoidable, but after second thoughts I think there's a legit trick in C++ to avoid the move. Will update.

gmagogsfm commented on torch/csrc/jit/runtime/script_profile.cpp:
```diff
@@ -0,0 +1,124 @@
+      }
+    }
+  }
+  datapoints_.clear();
```
This might be a surprise to users, maybe we should not clear datapoints.

zhxchen17 replied:
why? datapoints_ is just a buffer, and we only expose stats_ to users.

gmagogsfm commented on torch/csrc/jit/runtime/script_profile.cpp:
```diff
@@ -0,0 +1,124 @@
+}
+
+const ScriptProfile::Stats& ScriptProfile::dumpStats() {
+  TORCH_CHECK(!enabled_);
```
```suggestion
  TORCH_CHECK(!enabled_, "please disable XXX before dumping stats");
```

gmagogsfm commented on torch/csrc/jit/runtime/script_profile.cpp:
```diff
@@ -0,0 +1,124 @@
+  for (const auto& datapoint : datapoints_) {
+    if (const auto& source = datapoint.sourceRange.source()) {
+      if (auto fileLineCol = datapoint.sourceRange.file_line_col()) {
+        auto it = stats_.find(*source.get());
+        if (it == stats_.end()) {
+          it = stats_.emplace(SourceRef{source}, LineMap{}).first;
+        }
+        auto& stats = it->second[std::get<1>(*fileLineCol)];
```
Nit and optional:

Instead of performing a lookup and then conditionally insert new entry, you can unconditionally insert a new entry, which would either inserts a new entry and give you back the iterator pointing to it, or doesn't insert a new entry but still returns an iterator pointing to existing entry.

gmagogsfm reviewed (commented):
Looks great, just have a few nits here and there.

zhxchen17 commented on test/cpp/jit/test_script_profile.cpp:
```diff
@@ -0,0 +1,42 @@
+
+  ScriptProfile p;
+  p.enable();
+  {
+    auto g = profiling::InstructionSpan::tryMake(*node);
+    EXPECT_NE(g, c10::nullopt);
+  }
```
sure.

zhxchen17 commented on test/cpp/jit/test_script_profile.cpp:
```diff
@@ -0,0 +1,42 @@
+  torch::jit::parseIR(graph_string, &g);
+  auto source = std::make_shared<Source>(graph_string, "", 0);
+  auto node = *g.nodes().begin();
+  node->setSourceRange(SourceRange{source, 0, 0});
```
np

zhxchen17 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1392,6 +1393,19 @@ void initJITBindings(PyObject* module) {
           })
       .def(py::hash(py::self));
 
+  py::class_<InstructionStats>(m, "InstructionStats")
+      .def_property_readonly(
+          "count", [](const InstructionStats& self) { return self.count; })
+      .def_property_readonly("durationNs", [](const InstructionStats& self) {
+        return self.duration.count();
+      });
+
+  py::class_<ScriptProfile>(m, "_ScriptProfile")
+      .def(py::init<>())
+      .def("enable", [](ScriptProfile& p) { p.enable(); })
+      .def("disable", [](ScriptProfile& p) { p.disable(); })
+      .def("_dump_stats", [](ScriptProfile& p) { return p.dumpStats(); });
+
```
good point, let me have a look at TorchBind.

zhxchen17 commented on torch/csrc/jit/runtime/script_profile.cpp:
```diff
@@ -0,0 +1,124 @@
+  }
+}
+
+void ScriptProfile::addDatapoint(profiling::Datapoint datapoint) {
```
I'm not sure whether passing by shared_ptr is more performant, because we have one more heap allocation in that case. Meanwhile we have num_active_profiles more copies for the current approach, but my assumption is small copies like this should be super fast in modern architectures, and 90% of the cases num_active_profile == 1, so in common cases 1 allocation should take longer than 1 copy.

zhxchen17 committed:
addressed suggestions on "[jit] Implement ScriptProfile to collect instruction profiles."


Summary:
Introduces two main classes in C++ runtime:

ScriptProfile is the implementation for enalbing and disabling interpreter
profiling in C++. This should be only used from Python, and we will add
corresponding Python API in the next diff.

InstructionSpan is a utility class to instrument execution of each single
instruction. A start timestamp is recorded in the consturctor, and an end
timestamp is recorded in the destructor. During destruction, this will send
runtime data to all enabled ScriptProfile instances.

Test Plan:
build/bin/test_jit --gtest_filter='ScriptProfileTest.Basic'

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133579](https://our.internmc.facebook.com/intern/diff/D28133579)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133579).

zhxchen17 committed:
nits on "[jit] Implement ScriptProfile to collect instruction profiles."


Summary:
Introduces two main classes in C++ runtime:

ScriptProfile is the implementation for enalbing and disabling interpreter
profiling in C++. This should be only used from Python, and we will add
corresponding Python API in the next diff.

InstructionSpan is a utility class to instrument execution of each single
instruction. A start timestamp is recorded in the consturctor, and an end
timestamp is recorded in the destructor. During destruction, this will send
runtime data to all enabled ScriptProfile instances.

Test Plan:
build/bin/test_jit --gtest_filter='ScriptProfileTest.Basic'

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133579](https://our.internmc.facebook.com/intern/diff/D28133579)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133579).

zhxchen17 committed:
rebase on "[jit] Implement ScriptProfile to collect instruction profiles."


Summary:
Introduces two main classes in C++ runtime:

ScriptProfile is the implementation for enalbing and disabling interpreter
profiling in C++. This should be only used from Python, and we will add
corresponding Python API in the next diff.

InstructionSpan is a utility class to instrument execution of each single
instruction. A start timestamp is recorded in the consturctor, and an end
timestamp is recorded in the destructor. During destruction, this will send
runtime data to all enabled ScriptProfile instances.

Test Plan:
build/bin/test_jit --gtest_filter='ScriptProfileTest.Basic'

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133579](https://our.internmc.facebook.com/intern/diff/D28133579)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133579).

zhxchen17 committed:
Update on "[jit] Implement ScriptProfile to collect instruction profiles."


Summary:
Introduces two main classes in C++ runtime:

ScriptProfile is the implementation for enalbing and disabling interpreter
profiling in C++. This should be only used from Python, and we will add
corresponding Python API in the next diff.

InstructionSpan is a utility class to instrument execution of each single
instruction. A start timestamp is recorded in the consturctor, and an end
timestamp is recorded in the destructor. During destruction, this will send
runtime data to all enabled ScriptProfile instances.

Test Plan:
build/bin/test_jit --gtest_filter='ScriptProfileTest.Basic'

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133579](https://our.internmc.facebook.com/intern/diff/D28133579)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit] Implement ScriptProfile to collect instruction profiles."


Summary:
Introduces two main classes in C++ runtime:

ScriptProfile is the implementation for enalbing and disabling interpreter
profiling in C++. This should be only used from Python, and we will add
corresponding Python API in the next diff.

InstructionSpan is a utility class to instrument execution of each single
instruction. A start timestamp is recorded in the consturctor, and an end
timestamp is recorded in the destructor. During destruction, this will send
runtime data to all enabled ScriptProfile instances.

Test Plan:
build/bin/test_jit --gtest_filter='ScriptProfileTest.Basic'

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133579](https://our.internmc.facebook.com/intern/diff/D28133579)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133579).

gmagogsfm commented on torch/csrc/jit/runtime/script_profile.cpp:
```diff
@@ -0,0 +1,125 @@
+
+void ScriptProfile::addDatapoint(
+    std::shared_ptr<profiling::Datapoint> datapoint) {
+  TORCH_CHECK(enabled_, "Cannot only add datapoint to disabled profilers.");
```
```suggestion
  TORCH_CHECK(enabled_, "Cannot add datapoint to disabled profilers.");
```

zhxchen17 committed:
rebase on "[jit] Implement ScriptProfile to collect instruction profiles."


Summary:
Introduces two main classes in C++ runtime:

ScriptProfile is the implementation for enalbing and disabling interpreter
profiling in C++. This should be only used from Python, and we will add
corresponding Python API in the next diff.

InstructionSpan is a utility class to instrument execution of each single
instruction. A start timestamp is recorded in the consturctor, and an end
timestamp is recorded in the destructor. During destruction, this will send
runtime data to all enabled ScriptProfile instances.

Test Plan:
build/bin/test_jit --gtest_filter='ScriptProfileTest.Basic'

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133579](https://our.internmc.facebook.com/intern/diff/D28133579)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133579).

ezyang commented:
This diff broke torchdeploy tests: https://app.circleci.com/pipelines/github/pytorch/pytorch/325564/workflows/ddfae9dc-80b1-4514-98eb-9932787f2d85/jobs/13645042

cc @wconstab

wconstab commented:
I'm not familiar with torch::class_, some background on that would help me to explain this better.  I can't seem to find where torch::class_ is defined, a pointer would help.

Best guess of what's going wrong is, torch::deploy loads multiple copies of the python bindings, and in your case they now contain these torch::class_ instances, which apparently register themselves with some global in libtorch.  In cases where there are multiple torchdeploy interpreters, the first registration would work but subsequent ones complain due to the first one already being registered.  

A quick search shows there are some other torch::class_ usages, but not many, and i'm not sure if yours is the first one that appears in bindings that are a part of torchdeploy.  If we need to support this in torchdeploy, we should sync up and figure out how to do it.  Alternatively, you could ifdef these bindings to not be part of torchdeploy for now, or you could possibly make them py::class_ instead.

gmagogsfm commented:
> I'm not familiar with torch::class_, some background on that would help me to explain this better. I can't seem to find where torch::class_ is defined, a pointer would help.
> 
> Best guess of what's going wrong is, torch::deploy loads multiple copies of the python bindings, and in your case they now contain these torch::class_ instances, which apparently register themselves with some global in libtorch. In cases where there are multiple torchdeploy interpreters, the first registration would work but subsequent ones complain due to the first one already being registered.
> 
> A quick search shows there are some other torch::class_ usages, but not many, and i'm not sure if yours is the first one that appears in bindings that are a part of torchdeploy. If we need to support this in torchdeploy, we should sync up and figure out how to do it. Alternatively, you could ifdef these bindings to not be part of torchdeploy for now, or you could possibly make them py::class_ instead.

`torch::class_` is basically TorchBind, the main way to extend TorchScript with C++ extensions. There is a global registry that keeps track of all registered classes/methods. It is not clear to me why multiple Python instances brought up by `torch::deploy` would share global states instead of using completely separate memory spaces. Could you clarify?

`py::class_` is unfortunately not usable here because it is not compilable under TorchScript.

wconstab commented:
>torch::class_ is basically TorchBind, 
ok, i see, we'll have to figure this out then.

>why multiple Python instances brought up by torch::deploy would share global states
if the global states are part of the _python binding code_ then they would be replicated in each interpreter and have no issue.  But, torch::deploy doesn't replicate _all_ of libtorch, only the python bindings.  It shares the rest of libtorch, which is what enables sharing tensor data instead of copying it, across interpreters.  Hence, a global registry here probably won't work anymore, but i'm not sure what it'd take to make the registry 'interpreter-local'.

gmagogsfm commented:
> > torch::class_ is basically TorchBind,
> 
> ok, i see, we'll have to figure this out then.
> 
> > why multiple Python instances brought up by torch::deploy would share global states
> 
> if the global states are part of the _python binding code_ then they would be replicated in each interpreter and have no issue. But, torch::deploy doesn't replicate _all_ of libtorch, only the python bindings. It shares the rest of libtorch, which is what enables sharing tensor data instead of copying it, across interpreters. Hence, a global registry here probably won't work anymore, but i'm not sure what it'd take to make the registry 'interpreter-local'.

It sounds like making the registry "interpreter-local" is the goal here. I think that's very doable, in fact, should be very similar to PyBind case (I don't know how PyBind is replicated in torch::deploy though).

wconstab commented:
following up after offline chat-

it now sounds like the goal is the opposite- keep the class registry as a global, but don't call it multiple times.  We realized that the registry is just a mapping from string (python fqn of module) to c++ torchbind class, so it doesn't matter which python interpreter registers the name as long as one does.  

We also realized that the torch::class_ registrations probably don't need to be in the torch_python library which is replicated, so rather than finding a way to not do multiple registrations or allow multiple registrations, we can just try moving these new torch::class_  instances to a different cpp file that is part of libtorch not libtorch_python.

zhxchen17 commented:
just FYI: followup diff @ #58783
